File: c/linkedlist/linkedlist_cycle_ii.c
// #142 找到环形链表入口
#include <stdio.h>

struct Node
{
    struct Node *next;
    int val;
};

struct Node *entrance(struct Node *head)
{
    struct Node *slow = head, *fast = head;

    while (fast != NULL && fast->next != NULL)
    {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast)
        {
            while (head != slow)
            {
                head = head->next;
                slow = slow->next;
            }

            return head;
        }
    }

    return NULL;
}

File: go/linkedlist/linkedlist_cycle_ii.go
package linkedlist

// #142 寻找环的入口

type Node struct {
	Val  int
	Next *Node
}

// 2*(x+y) = x+y+n(y+z)
// x+y = n*(y+z)
// x = (n-1)(y+z)+z ---> x = z
func detectCycle(head *Node) *Node {
	fast, slow := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			for head != fast {
				head = head.Next
				fast = fast.Next
			}

			return head
		}
	}

	return nil
}
